### PR TITLE
fix:  skip codehash check when the code has been deleted in  the evm state

### PR DESCRIPTION
**Pull request by JayT106**
Closes: #1319

just ignores the codehash check when the evm account code is empty.

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/genesis.go [hunk 1]
```diff
@@ -94,3 +98,19 @@ func ExportGenesis(ctx sdk.Context, k *keeper.Keeper, ak types.AccountKeeper) *t
...
 	}
 }
+
+var (
+	emptyCodeHash  = crypto.Keccak256(nil)
+	patchCodeHash1 = common.HexToHash("0x1d93f60f105899172f7255c030301c3af4564edd4a48577dbdc448aec7ddb0ac").Bytes()
+	patchCodeHash2 = common.HexToHash("0x6dbb3be328225977ada143a45a62c99ace929f536b75a27c09b6a09187dc70b0").Bytes()
+)
+
+// checkCodeHash return false if the evm state code was been deleted, see ethermint PR#1234
```
Comment (facs95) [inline x/evm/genesis.go:108]:
```suggestion
// checkCodeHash return false if the evm state code has been deleted, see ethermint PR#1234
```

File: x/evm/genesis.go [hunk 2]
```diff
@@ -94,3 +98,19 @@ func ExportGenesis(ctx sdk.Context, k *keeper.Keeper, ak types.AccountKeeper) *t
 		Params:   k.GetParams(ctx),
 	}
 }
+
+var (
+	emptyCodeHash  = crypto.Keccak256(nil)
+	patchCodeHash1 = common.HexToHash("0x1d93f60f105899172f7255c030301c3af4564edd4a48577dbdc448aec7ddb0ac").Bytes()
```
Comment (facs95) [inline x/evm/genesis.go:104]: Sorry, not following what these two code hashes represent. Can you please explain?

Comment (JayT106) [reply x/evm/genesis.go:104]: `0x6dbb3be328225977ada143a45a62c99ace929f536b75a27c09b6a09187dc70b0` is the codehash of the code `0x6d4946c0e9f43f4dee607b0ef1fa1c3314585733ff`, it contains `selfdestruct`
I am trying to dumping the code for the other one, I believe it's similar case, will update it in the PR.

File: x/evm/genesis.go [hunk 3]
```diff
@@ -94,3 +98,19 @@ func ExportGenesis(ctx sdk.Context, k *keeper.Keeper, ak types.AccountKeeper) *t
...
+	patchCodeHash2 = common.HexToHash("0x6dbb3be328225977ada143a45a62c99ace929f536b75a27c09b6a09187dc70b0").Bytes()
+)
+
+// checkCodeHash return false if the evm state code has been deleted, see ethermint PR#1234
+func checkCodeHash(evmCodeHash []byte, accountCodeHash []byte) bool {
```
Comment (yihuang) [inline x/evm/genesis.go:109]: Is this function works like `isCodeHashMatch`?

**Review by yihuang — APPROVED** [review]
LGTM

File: x/evm/genesis.go [hunk 4]
```diff
@@ -48,11 +48,14 @@ func InitGenesis(
 				),
 			)
 		}
-
 		code := common.Hex2Bytes(account.Code)
 		codeHash := crypto.Keccak256Hash(code)
-		if !bytes.Equal(ethAcct.GetCodeHash().Bytes(), codeHash.Bytes()) {
-			panic("code don't match codeHash")
+
+		// we ignore the empty Code hash checking, see ethermint PR#1234
```
Comment (fedekunze) [inline x/evm/genesis.go:54]:
```suggestion
		// ignore the empty code hash check. See Ethermint Security Advisory https://github.com/evmos/ethermint/security/advisories/GHSA-f92v-grc2-w2fg
```

**Review by fedekunze — APPROVED** [review]
LGTM. Can you also add a unit test case with the empty code hash?

File: x/evm/genesis_test.go [hunk 5]
```diff
@@ -96,6 +96,24 @@ func (suite *EvmTestSuite) TestInitGenesis() {
 			},
 			true,
 		},
+		{
+			"ignore empty account code checking",
+			func() {
+				acc := suite.app.AccountKeeper.NewAccountWithAddress(suite.ctx, address.Bytes())
+
+				suite.app.AccountKeeper.SetAccount(suite.ctx, acc)
```
Comment (yihuang) [inline x/evm/genesis_test.go:104]: Set a non-empty code hash to the account?

Comment (JayT106) [reply x/evm/genesis_test.go:104]: Yes, the codehash in the account keeper will be an empty code hash in this test case. However, when we set the Account code to empty in the evm state. No matter whether the codehash is empty or not, it always passes the check.

Comment (yihuang) [reply x/evm/genesis_test.go:104]: Because empty code hash also pass the test even without this change, right? So a non-empty one can demonstrate the change works?

Comment (JayT106) [reply x/evm/genesis_test.go:104]: I will try to add another one with non-empty hash case

**Comment by yihuang** [comment]
there's conflicts need to resolve?

**Comment by yihuang** [comment]
LGTM, can we merge this now @fedekunze ?

**Comment by yihuang** [comment]
LGTM, can we merge this now @fedekunze ?

**Comment by danburck** [comment]
@yihuang isn't the test with empty code hash still missing? 

```
emptyCodeHash := crypto.Keccak256(nil)
codeHash := common.BytesToHash(emptyCodeHash).String()
```

**Comment by yihuang** [comment]
> @yihuang isn't the test with empty code hash still missing?
> 
> ```
> emptyCodeHash := crypto.Keccak256(nil)
> codeHash := common.BytesToHash(emptyCodeHash).String()
> ```

Which specific case do you have in mind, the default account has the empty code hash, so the previous test cases tested that.
